code/core/src/preview/preview-navigator.ts

```diff
@@ -0,0 +1,193 @@
+  const index = (await (await fetch('/index.json')).json()) as StoryIndex;
+
+  const currentEntryId = url.searchParams.get('id');
+  if (!currentEntryId) {
+    const firstEntry = Object.values(index.entries)[0];
+    if (firstEntry) {
+      url.searchParams.set('id', firstEntry.id);
+      url.searchParams.set('viewMode', firstEntry.type);
+      window.location.href = url.toString();
+    }
+    return;
+  }
+  setupPreviewNavigator(index, currentEntryId);
+}
+
+export const setupPreviewNavigator = async (index: StoryIndex, currentEntryId: string) => {
+  const tree: BranchNode = { title: '', entries: {}, isActive: true };
+  for (const entry of Object.values(index.entries)) {
+    const titleParts = entry.title.split('/');
+
+    let currentNode = tree;
+    for (const titlePart of titleParts) {
+      if (!currentNode.entries) {
+        currentNode.entries = {};
+      }
+      if (!currentNode.entries[titlePart]) {
+        currentNode.entries[titlePart] = {
+          title: titlePart,
+          isActive: currentEntryId === entry.id,
+          entries: {},
+        };
+      } else if (currentEntryId === entry.id) {
+        currentNode.entries[titlePart].isActive = true;
+      }
+      currentNode = currentNode.entries[titlePart] as BranchNode;
+    }
+    if (!currentNode.entries) {
+      currentNode.entries = {};
+    }
+    currentNode.entries[entry.name] = {
+      id: entry.id,
+      name: entry.name,
+      href: `?id=${entry.id}&viewMode=${entry.type}&navigator=true`,
+      isActive: currentEntryId === entry.id,
+    };
+  }
+
+  const createHtmlForNode = (node: BranchNode | LeafNode): string => {
+    if ('entries' in node && 'title' in node) {
+      const branchNode = node as BranchNode;
+      return `
+        <details${branchNode.isActive ? ' open' : ''}>
+          <summary class="sb-navigator-title">
+            ${branchNode.title}
+          </summary>
+          <ul class="sb-navigator-entries" aria-label="${branchNode.title}">
+            ${Object.values(branchNode.entries).map(createHtmlForNode).join('')}
+          </ul>
+        </details>
+      `;
+    }
+
+    const leafNode = node as LeafNode;
+    return `
+      <li class="sb-navigator-story-item">
+        <a href="${leafNode.href}" 
+           class="sb-navigator-story-link${leafNode.isActive ? ' active' : ''}" 
+           aria-current="${leafNode.isActive ? 'location' : 'false'}">${leafNode.name}</a>
+      </li>
+    `;
+  };
+
+  const navItems = Object.values(tree.entries).map(createHtmlForNode).join('');
+
+  const nav = document.createElement('nav');
+  nav.className = 'sb-navigator-nav';
+  nav.setAttribute('role', 'navigation');
+  nav.setAttribute('aria-label', 'Story navigation');
+  nav.innerHTML = `
+    <ul class="sb-navigator-list">${navItems}</ul>
+  `;
+
+  document.body.insertBefore(nav, document.body.firstChild);
+
+  const style = document.createElement('style');
+  style.textContent = `
```

Comment:
I'm not a huge fan of this wall of CSS string here, it feels brittle. However trying to extract this to a CSS file and dynamically adding it to the head felt like overkill